rfcs/0124-worker-manager.md

```diff
@@ -0,0 +1,271 @@
+configuration or worker.  The data backing this information will be received
+from the Queue over RabbitMQ.  The worker manager will track outcomes.  The
+only information from the Queue stored in this service will be the task run
+outcome, task id and run id.  All other information will be available through
+the Queue.
+
+### Worker Configuration
+The worker manager will provide facilities for configuring Workers.  There will
+be a set of Taskcluster APIs for creating and managing worker configurations.
+Each worker configuration will specify global options like the maximum global
+capacity units, the scaling ratio and the list of worker types for which this
+worker configuration can be used.
+
+Worker configurations will also have per-providers options.  Each provider's
+options section will be able to override global options, like max capacity.
+This would be used to say "We want to have 1000 Capacity, but no more than 100
+from hardware".  Each provider a worker configuration can use would also have a
+section for provider-specific configuration.  This would store things like
+which worker image to use.  A more concrete example is that this section would
+store the AMI to use for an EC2 Worker.
+
+All levels of worker type configurations (e.g. global, per worker-type,
+per-provider) will support a `disabled: true` which will disable that section
+of the configuration.  The least specific disable will take priority.  This
+value will only block creation of new resources, not disable management of
+existing resources.
+
+Each edit of a worker-configuration will only be atomic to the
+worker-configuration level.
+
+### Multiple Worker Types for one Worker Configuration
+There are a lot of worker types which have identical configuration other than
+their worker type.  In order to support having more than one worker type for a
+given worker configuration, the worker configuration will have a list of worker
+types for which it can provide capacity for.  This will be a list in the
+configuration document.  Each entry in the list is either a string or an object
+which overrides global worker configuration options like max capacity for that
+specific worker type.  For configuration which is a range or value, the most
+restrictive will be used.
+
+Each worker type must be globally unique to a single worker configuration.
+
+### Documentation and Payload Schemata
+Each worker type and worker configuration might have documentation and payloads
+associated with them.
+
+The schema will be exposed as any other value in the worker configuration to
+the JSON document as an option like any other, but will be treated as an opaque
+JSON-Schema document.  If overridden in a more specific configuration section,
+the most specific complete document will be used.  Accessing the payload schema
+by worker type will be supported.
+
+Documentation will be supported through a second API endpoint.  Documentation
```

Comment:
What if we supported the [taskcluster-lib-docs format](https://docs.taskcluster.net/docs/reference/libraries/taskcluster-lib-docs/docs/format) here. That way we can support images/diagrams/multiple pages etc. and also taskcluster-docs already knows how to work with it and display it.

Reply:
Then we could show these docs in the docs site instead of going through this service to render them. (Which is what I think is being described here?)

Reply:
I changed to saying a tarball of this format.  my intention is that this should allow for future integration but that we shouldn't specify this as a part of a first version.  Let me know what you think!

Reply:
I like this a lot. @djmitche, iirc you had an issue with my tarball idea though? Should I not have suggested this?

Reply:
I suspect that tarball format will fade away in favor of something else once we're not shipping tarballs to S3 anymore, and it does seem a bit odd for an API to store tarballs.  Also, this would be the first "runtime configuration" to show up in the docs, and we have no precedent for that (and it will not be possible until tc-web includes docs - even then I'm not sure how it will work).  Rather than leave a big question-mark on this issue, I think we should just propose a markdown blob and plan to change it later if necessary.

I suspect that the multi-page stuff, and images, would be in one place -- the generic-worker documentation.  I think worker configuration documentation would mostly just link to that, perhaps to sub-pages.  And those URLs can be relative (`/docs/reference/workers/generic-worker/docs/docker-tasks`).